### DISCUSSION TITLE
Tasmota Smart Meter Interface Timeout detection

### DISCUSSION THREAD

**Discussion by Jupp2** [arendst/Tasmota#20790] [Q&A] [answered]
Hi guys,

I have a running smart meter interface so far. I would like to detect when there is no sml data received in Tasmota. Currently when pulling off the reading head from the smart meter, the script examples and my script still publish the last known wattage to mqtt. Is there a way to detect no new valid sml data was received within a defined timeout like e.g. 10s. 
Tried things like upd[sml]>0 which did not work.
Currently as workaround I check if no change in wattage occured over 120s, however this is not as fast as desired and sometimes generates "false-alarms".

>D
watt=0
cnt=0
sumneg=0
sumpos=0
NegWs=0
PosWs=0
PavgW=0
PmaxW=0
PminW=0
PdiffW=0
>B
tper=120
->sensor53 r
=>publish tele/%topic%/tper %tper% 
>S
;Writes the value of Descriptorline 3 to watt
watt=sml[3]

if watt>PmaxW
then
PmaxW=watt
endif
if watt<PminW
then
PminW=watt
endif
if upsecs%tper==0
then
if cnt>0
then
NegWs=sumneg*tper/cnt
PosWs=sumpos*tper/cnt
PavgW=(sumpos+sumneg)/cnt
else
NegWs=0
PosWs=0
PavgW=0
endif
cnt=0
sumpos=0
sumneg=0
PdiffW=PmaxW-PminW
=>publish tele/%topic%/SML_PminW %PminW% 
=>publish tele/%topic%/SML_PmaxW %PmaxW%
=>publish tele/%topic%/SML_NegWs %NegWs%
=>publish tele/%topic%/SML_PosWs %PosWs%
=>publish tele/%topic%/SML_movavW %PavgW%
=>publish tele/%topic%/SML_PdiffW %PdiffW% 
PminW=watt
PmaxW=watt
endif
if watt>=0
then
sumpos=sumpos+watt
cnt=cnt+1
else
if watt<0
then
sumneg=sumneg+watt
cnt=cnt+1
endif
endif
>M 1
+1,3,s,16,9600,SML
1,77070100600100ff@#,Server ID,,server_id,0
1,77070100010800ff@1000,Energy,kWh,total_kwh,4
1,=so1,00010800,65,11,65,11,00100700
1,77070100100700ff@1,Power,W,curr_w,0
#

**Comment by gemu2015** [comment, marked as answer]
smlv[x]  is >0 when a value was received (see scripting docs)
example:

```
>D
valid=0

>B
=>sensor95 r

>S
; check every 30 seconds
if upsecs%30==0 {
       ; Check if decode line 1 (Verbrauch) was updated
	valid=smlv[1]
	if valid==0 {
		; no update in the last 30 seconds
		; do whatever you like here 
	} else {
		; this resets all valid flags
		valid=smlv[0]
	}
}

>M 1
+1,3,o,0,9600,OBIS
1,1-0:1.8.1*255(@1,Verbrauch,KWh,Total_in,4
1,1-0:2.8.1*255(@1,Einspeisung,KWh,Total_out,4
1,=d 2 10 @1,Aktueller Verbrauch,W,Power_curr,0
1,1-0:0.0.0*255(@#),Zähler Nr,, Meter_number,0
#
```

**Reply by Jupp2** [reply]
Thanks a lot! This works for me sending a timeout warning after 10s:

>D
valid=0
t:timeout=-1
alive=0

>S
;Writes the value of Descriptorline 3 to watt
watt=sml[3]
valid=smlv[3]
	if valid==1
	then
		; if timeout was expired and now is running again, publish alive message 
		if timeout<0
		then
			alive=1
			=>publish tele/%topic%/SML_alive %0alive%
		endif
		; updated power data resets countdown timer to 10 seconds
		timeout=10
	endif

; timeout expired, publish not-alive message 
if timeout==0
then 
	alive=0
	=>publish tele/%topic%/SML_alive %0alive%
	;stop timeout timer
	timeout=-1
endif

; reset all SML valid flags
valid=smlv[0]
